### PR TITLE
CHANGELOG/documentation updates for 0.23

### PR DESCRIPTION
**Pull request by bendk**
- Added CHANGELOG sections to describe the new crate structure and breaking changes from PR #1374
- Updated the manual sections describing how to set up `uniffi-bindgen` binaries.  In particular, we're now suggesting naming the binaries `uniffi-bindgen` instead of `run-uniffi-bindgen`.  The "run" part seemed unnecessary to me, especially since `uniffi-bindgen` no longer provides a standalone binary.

My plan is to make a new UniFFI release once this is approved and merged.

### CODE REVIEW

File: CHANGELOG.md
Comment (badboy) [inline CHANGELOG.md:13]:
```suggestion
- `uniffi_bindgen` no longer provides a standalone binary.  Having a standalone binary resulted in version mismatches when the `uniffi` version specified in `Cargo.toml` didn't match the `uniffi_bindgen` version installed on the system.  Read [The foreign language bindings](https://mozilla.github.io/uniffi-rs/tutorial/foreign_language_bindings.html) section of the manual for how to set up a `uniffi-bindgen` binary that's local to your workspace.
```

File: CHANGELOG.md
Comment (badboy) [inline CHANGELOG.md:20]: Should we update our own docs at `docs/release-process.md` on how to release individual crates?

**Review by badboy — CHANGES_REQUESTED** [review]

File: CHANGELOG.md
Comment (bendk) [reply CHANGELOG.md:20]: Yeah, but I actually wasn't sure how the new release process would work now.  I was thinking we could run through a few releases and update the docs as we went.  I'm not really sure though, WDYT?

Comment (mhammond) [reply CHANGELOG.md:20]: We might as well make a best-effort attempt - I assume you will be making the release, so maybe note what you expect to do, then we can tweak as we go?  (ie, if I had to make it, I wouldn't really know where to start)

Comment (bendk) [reply CHANGELOG.md:20]: That makes sense to me, just pushed a commit with my attempt at this.

File: docs/release-process.md
Comment (bendk) [inline docs/release-process.md:29]: Having to maintain two separate changelogs doesn't seem great, but I can't think of a better way to handle this.

AFAICT, the changelogs are not published anywhere they're only available by browsing github.  If this is true, it's nice because I don't need to update any automation.  But maybe we should consider including the changelog in the docs.

**Comment by bendk** [comment]
Pushed some updates based on the feature gate changes.  I think this is basically ready, except for one thing:  I added a "how to migrate" section in the changelog, but I wonder if there's a better way to get this info out to UniFFI users since the changelog isn't very visible at this point.  I'm not sure what's a better place though?  The manual?  The README.md file?

File: docs/manual/src/tutorial/foreign_language_bindings.md
Comment (mhammond) [inline docs/manual/src/tutorial/foreign_language_bindings.md:17]: The name isn't important, right? Maybe a comment saying something like "# whatever name makes sense for your project can be used, but that the rest of this tutorial assumes `uniffi-bindgen`"?

File: docs/manual/src/tutorial/foreign_language_bindings.md
Comment (mhammond) [inline docs/manual/src/tutorial/foreign_language_bindings.md:36]: typo - /seciton/

File: docs/release-process.md
Comment (mhammond) [inline docs/release-process.md:24]: this sentence reads a little odd - "users `uniffi` of the UniFFI crate." - do you mean something like "users of the top-level `uniffi` crate"?

File: docs/release-process.md
Comment (mhammond) [reply docs/release-process.md:29]: I don't understand why we need a new CHANGELOG, and don't understand how I'd actually use both (ie, *every* change to the sub-crates potentially impacts users of the top-level crate, as many features will be released without a top-level version bump. I assume this has something to do with communicating how breaking a breaking change is?

File: docs/release-process.md
Comment (mhammond) [inline docs/release-process.md:34]: Ah, for some reason I was assuming we'd only bump version numbers for sub-crates which changed, so those numbers would start diverging - but given the above, that's not correct, right? I wonder if that's worth spelling out somewhere?

File: docs/release-process.md
Comment (mhammond) [inline docs/release-process.md:67]: Is there a step missing here to update `uniffi/Cargo.toml` with the new version numbers of the sub-crates - there's also a bit of a chicken-and-egg situation here - don't the sub-crates need to be on crates.io before we can release the top-level crate?

**Review by mhammond — COMMENTED** [review]
Thanks for the update! It did raise some more questions for me (which is good I suppose :)

File: docs/release-process.md
Comment (bendk) [reply docs/release-process.md:29]: Some changes will only affect the sub crates.  Here's some examples from the current CHANGELOG:
 -  The `guess_crate_root` function is now public
- `uniffi_bindgen`: Renamed `FFIArgument`, `FFIFunction` and `FFIType` to `FfiArgument`, `FfiFunction` and `FfiType` (this was a breaking change).

The issue you mentioned was also on my mind: a change could be breaking for sub-crate users, but non-breaking for top-level users.  However, when I looked through the changelog, I didn't see any examples of those.

The last issue was version numbers.  The version numbers for `uniffi` and the sub-crates are going to start to diverge, which makes the changelog tricky.  I don't know how we would setup the `cargo release` template to move changes from the unreleased section, if there are 2 different version numbers involved.

File: docs/release-process.md
Comment (bendk) [reply docs/release-process.md:67]: I think that `cargo release-backend-crates` should do that, although I'm not sure.  I'll verify that when I do the 0.23 release.  I added a warning to that section.

File: CHANGELOG.md
Comment (mhammond) [inline CHANGELOG.md:23]: I was wondering whether we could move this section to the top of the file above `<!-- next-header -->` to make it somewhat "sticky", changing the title to something like "Migrating from UniFFI 0.22"? It would probably need an extra introductory para to say something like "UniFFI 0.23 had significant changes to the versioning scheme, here's the info you need to upgrade" too.

My vague concern is that we start leveraging the new versioning scheme to make new releases far more often just to get new features someone wants to leverage. So a UniFFI user might not actually attempt to upgrade until we've made 5 new releases (with `uniffi` still on 0.23.x but the sub-crates up to 0.27.0), so this info might get lost in the noise.

**Comment by bendk** [comment]
Pushed some changes after talking CHANGELOGs with @mhammond:
  - Moved the `Upgrading to 0.23` section to the top of the changelog.  I agree it's good to keep that visible for a while
  - Removed `uniffi/CHANGELOG.md`.  Mark convinced me that it's better to just keep one changelog and label what affects "regular" users vs what affects external bindings authors.  How do we do that?  I'm not totally sure, I think we should just wing it and discuss the larger issue of changelogs in #1458
  - Updated the `release.toml` scripts to hopefully update that changelog correctly when we do the release.  I'll definitely monitor this when it's release time though.

WDYT, can we merge this one and start the `0.23` release?

**Review by mhammond — APPROVED** [review]
Thanks for working through this Ben! Probably worth waiting for @badboy's approval too though.

**Comment by badboy** [comment]
> Thanks for working through this Ben! Probably worth waiting for @badboy's approval too though.

on it today

**Review by badboy — APPROVED** [review]
minor nits, but otherwise good to go.

File: CHANGELOG.md
Comment (bendk) [reply CHANGELOG.md:6]: @mhammond suggested making it sticky for a while, in case a user misses several releases and then tries to upgrade.  This makes sense for about 6-12 months, then we can move it back down.  That said, I don't have strong feelings about this one.

Comment (badboy) [reply CHANGELOG.md:6]: Ah, didn't see that. Yeah, I'm fine with that.

**Pull request merged**
